qutip/solver/integrator.py

```diff
@@ -0,0 +1,262 @@
+""" `Integrator`: ODE solver wrapper to use in qutip's Solver """
+import numpy as np
+
+__all__ = ['Integrator', 'IntegratorException', 'add_integrator',
+           'sesolve_integrators', 'mesolve_integrators', 'mcsolve_integrators']
+
+
+class IntegratorException(Exception):
+    """Error from the ODE solver being unable to integrate with the given
+    parameters.
+
+    Example
+    -------
+    - The solver cannot reach the desired tolerance within the maximum number
+    of steps.
+    - The step needed to be within desired tolerance is too small.
+    """
+
+
+class Integrator:
+    """
+    A wrapper around ODE solvers.
+    It ensures a common interface for Solver usage.
+    It takes and return states as :class:`qutip.core.data.Data`, it may return
+    a different data-type than the input type.
+
+    Parameters
+    ----------
+    system: qutip.QobjEvo
+        Quantum system in which states evolve.
+
+    options: qutip.SolverOdeOptions
+        Options for the solver.
+
+    Class Attributes
+    ----------------
+    supports_blackbox : bool
+        If True, then the integrator calls only ``system.matmul``,
+        ``system.matmul_data``, ``system.expect``, ``system.expect_data`` and
+        ``isconstant``, ``isoper`` or ``issuper``. This allows the solver using
+        the integrator to modify the system in creative ways. In particular,
+        the solver may modify the system depending on *both* the time ``t``
+        *and* the current ``state`` the system is being applied to.
+
+        If the integrator calls any other methods, set to False.
+
+    supports_time_dependent : bool
+        If True, then the integrator supports time dependent systems. If False,
+        ``supports_blackbox`` should usually be ``False`` too.
+
+    integrator_options : dict
+        A dictionary of options used by the integrator and their default
+        values. Once initiated, ``self.options`` will be a dict with the same
+        keys, not the full options object passed to the solver. Options' keys
+        included here will be supported by the :cls:SolverOdeOptions.
+    """
+    # Used options in qutip.SolverOdeOptions
+    integrator_options = {}
+    # Can evolve time dependent system
+    support_time_dependant = None
+    # Use the QobjEvo's matmul_data method as the driving function
+    supports_blackbox = None
+    name = ""
```

Comment:
```suggestion
    # The name of the integrator
    name = None
```
I suggest we set this to `None` so that it is an error to not specify a name for the solver.